[14.0][ADD] delivery_purchase_label

TDu opened the pull request:
Print delivery labels for dropshipping purchase order. Useful when then vendor that will process the order does not have the capabilities to print the transporter labels needed for the delivery.

TDu committed:
Add delivery_purchase_label

Print delivery labels for dropshipping purchase order.
Useful when then vendor that will process the order does not have the
capabilities to print the transporter labels needed for the delivery.

henrybackman commented on delivery_purchase_label/readme/CONFIGURE.rst:
```diff
@@ -0,0 +1,8 @@
+
+In "Inventory > Configuration > Delivery > Shipping Methods", a new field
+`Operation Type for Sending Labels`
+allows to select the opeartion type to use for the stock transfer that will
```
```suggestion
allows to select the operation type to use for the stock transfer that will
```

henrybackman commented on delivery_purchase_label/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,15 @@
+A new operation type is created by the module and used for the transfer.
+All attachment (should be only the labels) on the transfer are then attached
+to the email send to the vendor (rfq or po).
+When the purhcase order is still in draft the label are regenerated everytime
```
```suggestion
When the purchase order is still in draft the labels are regenerated everytime
```

henrybackman commented on delivery_purchase_label/models/purchase_order.py:
```diff
@@ -0,0 +1,93 @@
+
+    def action_rfq_send(self):
+        self.ensure_one()
+        if self._states:
```
It's not obvious to me why the check on `_states` here. What does it mean?

TDu replied:
Good catch this was a left over from refactoring

henrybackman commented on delivery_purchase_label/models/purchase_order.py:
```diff
@@ -0,0 +1,93 @@
+        """Create a transfer to generate the carrier labels."""
+        self.ensure_one()
+        if self.state not in self._states_to_generate_delivery_label():
+            return
```
Could this and the following `product.type` check be in `_is_valid_for_vendor_labels()`?

TDu replied:
yes it could, changed.

henrybackman commented on delivery_purchase_label/models/purchase_order.py:
```diff
@@ -0,0 +1,93 @@
+            for product in self.order_line.product_id
+        ):
+            return
+        # Find the carrier and that will be used
```
```suggestion
        # Find the carrier that will be used
```

henrybackman reviewed (commented):
A couple of small comments

jbaudoux commented on delivery_purchase_label/__manifest__.py:
```diff
@@ -0,0 +1,20 @@
+    "version": "14.0.1.0.0",
+    "category": "Delivery",
+    "website": "https://github.com/OCA/delivery-carrier",
+    "author": "Camptocamp, Odoo Community Association (OCA)",
```
```suggestion
    "author": "Camptocamp, BCIM, Odoo Community Association (OCA)",
```

vuwnevska commented on delivery_purchase_label/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,15 @@
+Print delivery labels for dropshipping purchase order.
+Useful when then vendor that will process the order does not have the
```
```suggestion
Useful when the vendor that will process the order does not have the
```

ivantodorovich commented on delivery_purchase_label/models/mail_compose_message.py:
```diff
@@ -0,0 +1,39 @@
+            in self.env["purchase.order"]._mail_templates_to_not_attach_labels()
+        ):
+            return
+        if isinstance(res_ids, int):
+            res_ids = [res_ids]
```
```suggestion
```

This doesn't seem necessary

given `self.env["purchase.order"].browse(1) == self.env["purchase.order"].browse([1])`

ivantodorovich reviewed (approved):
Thanks !

It seems there are quite some lines not covered by unit tests. It might be a good idea to add a few more cases to them :)

LGTM

jbaudoux commented on delivery_purchase_label/models/purchase_order.py:
```diff
@@ -0,0 +1,93 @@
+            "partner_id": self.dest_address_id.id,
+            "user_id": False,
+            "date": fields.Datetime.now(),
+            "origin": self.name,
```
or maybe the SO number ? To clarify
```suggestion
            "origin": self.origin,
```

mt-software-de replied:
Yes use self.origin this makes it much easier to generate the label out the purchae delivery label record. 
It represents then more or less a copy of the initial purchase order.

jbaudoux replied:
Don't we wanted both on label? If we concatenate the two doesn't it solve the label?
"%origin - %name" ?

mt-software-de replied:
ping @TDu

mt-software-de commented on delivery_purchase_label/models/purchase_order.py:
```diff
@@ -0,0 +1,93 @@
+            "mail.message_origin_link",
+            values={"self": picking, "origin": self},
+            subtype_id=self.env.ref("mail.mt_note").id,
+        )
```
This will create a new transfer on every rfq send. 
We should check if something is created or if the create delivery transfer is still equal to the lines in the purchase.order, otherwise cancel the delivery transfer a create a new one.

mt-software-de commented on delivery_purchase_label/models/purchase_order.py:
```diff
@@ -0,0 +1,93 @@
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    delivery_label_picking_id = fields.Many2one(
+        "stock.picking",
+        "Delivery Label Picking",
+        store=True,
+        readonly=True,
+    )
```
Could you please add this field to the view.

mt-software-de commented on delivery_purchase_label/models/purchase_order.py:
```diff
@@ -0,0 +1,93 @@
+        picking.sale_id = False
+        moves.sale_line_id = False
+        moves.purchase_line_id = False
+        picking.action_assign()
+        for move in moves:
```
The moves can change on action_assign
```suggestion
        picking.action_assign()
        moves = picking.move_lines
        for move in moves:
```

mt-software-de commented on delivery_purchase_label/data/stock_picking_type.xml:
```diff
@@ -0,0 +1,9 @@
+        <field name="code">internal</field>
+        <field name="sequence_code">LBL</field>
+        <field name="default_location_dest_id" ref="stock.stock_location_suppliers" />
+        <field name="default_location_src_id" ref="stock.stock_location_suppliers" />
```
Why is this suppliers? And not customers?
```suggestion
        <field name="default_location_src_id" ref="stock.stock_location_suppliers" />
```

mt-software-de commented on delivery_purchase_label/data/stock_picking_type.xml:
```diff
@@ -0,0 +1,9 @@
+        <field name="name">Dropship Carrier Label</field>
+        <field name="code">internal</field>
+        <field name="sequence_code">LBL</field>
+        <field name="default_location_dest_id" ref="stock.stock_location_suppliers" />
```
Why is there the suppliers location used?
And not customers?
And why is the code internal and not outgoing?
What are the reasons for that? Stock valuation?

jbaudoux replied:
It's on purpose. It's the supplier packing operation, so an internal operation done by the supplier. Goods are not yet sent.

mt-software-de replied:
Yes and it is good for valuation layer, because if it would be supplier -> customer is would be recognized as dropship and will create a valuation for it.

mt-software-de commented on delivery_purchase_label/models/purchase_order.py:
On cancel the created label picking should be as well cancelled.

jbaudoux replied:
seems tackled now

mt-software-de commented on delivery_purchase_label/models/delivery_carrier.py:
```diff
@@ -0,0 +1,12 @@
+class DeliveryCarrier(models.Model):
+    _inherit = "delivery.carrier"
+
+    purchase_label_picking_type = fields.Many2one(
+        string="Operation Type For Purchase Label", comodel_name="stock.picking.type"
+    )
```
Is this really needed to be set on the carrier?
Or is it enough to just simple configure the carrier on the partner?
And the for creating the purchase delivery label - just use the operation type from data xml?

This would result in less code and it would ensure that there is just one operation type for this matter. 

```suggestion
```

jbaudoux replied:
The feature needs to be enabled somewhere. On the carrier is the best place and configuring the operation type is also legit

mt-software-de replied:
yes the feature needs to be enabled, but it is already enabled on the partner where the carrier is set.

mt-software-de replied:
The background to this question is.
How do you find out which transfers are "purchase deliver label" transfers. 

The simplest way would be to have just one operation type for all of those transfers. This would mean no configuration on carrier just on the supplier. 

With the current implementation there could be a lot types used for that. 
Which makes it a lot harder to find the right transfers. 

A other possible way would be to mark those operation types as delivery purchase label type.

mt-software-de commented on delivery_purchase_label/models/purchase_order.py:
```diff
@@ -0,0 +1,95 @@
+            return
+        # Find the carrier that will be used
+        carrier = self.partner_id.purchase_delivery_carrier_id
+        if not carrier.purchase_label_picking_type:
+            return
```
Is it needed to check the `purchase_label_picking_type` or can we assume that if the `purchase_delivery_carrier_id` is set on the partner we are just using the created operation type from the data file. 
```suggestion
        if not carrier:
            return
```

jbaudoux replied:
It's better to configure than assume anything

mt-software-de commented on delivery_purchase_label/models/purchase_order.py:
```diff
@@ -0,0 +1,95 @@
+
+    def _get_purchase_delivery_label_picking_value(self, carrier):
+        return {
+            "picking_type_id": carrier.purchase_label_picking_type.id,
```
```suggestion
            "picking_type_id": self.env.ref("delivery_purchase_label.picking_type_send_label"),
```

jbaudoux replied:
it is good as it is

mt-software-de commented on delivery_purchase_label/models/purchase_order.py:
```diff
@@ -0,0 +1,132 @@
+            )
+        if self._is_picking_label_uptodate():
+            return
+        order = self.with_company(self.company_id)
+        # Create and process the transer to send the labels
+        values = order._get_purchase_delivery_label_picking_value(carrier)
+        picking = self.env["stock.picking"].with_user(SUPERUSER_ID).create(values)
+        moves = order.order_line._create_stock_moves(picking)
+        moves.location_id = picking.location_id
+        moves.location_dest_id = picking.location_dest_id
+        # Remove the link on the sale and purchase
+        # To not impact the delivered quantity on them
+        picking.sale_id = False
+        moves.sale_line_id = False
+        moves.purchase_line_id = False
+        picking.action_assign()
+        # Moves can change on action assign
+        moves = picking.move_lines
+        for move in moves:
+            move.quantity_done = move.product_uom_qty
```
Could this please move into an own method.
Including also line 113 `picking._action_done`

mt-software-de commented on delivery_purchase_label/models/purchase_order.py:
```diff
@@ -0,0 +1,132 @@
+        moves = picking.move_lines
+        for move in moves:
+            move.quantity_done = move.product_uom_qty
+        if order.delivery_label_picking_id:
+            # Canceling the previous picking.
+            pick = order.delivery_label_picking_id
+            # Using wirte to by pass internal checks
+            # Should not be a problem because this is a fake
+            # move (Vendor to Vendor)
+            pick.move_line_ids.write({"state": "cancel"})
+            pick.move_lines.write({"state": "cancel"})
+            pick.write({"state": "cancel"})
```
Could this also move into an own method. 
To make it hookable and please also call `pick.cancel_shipment`

jbaudoux replied:
Yes, indeed, make a method for canceling the delivery label picking.
You need to call this also on PO cancellation as mentioned by MT

jbaudoux commented on delivery_purchase_label/models/purchase_order.py:
```diff
@@ -0,0 +1,132 @@
+        if not self._is_valid_for_vendor_labels():
+            return
+        # Find the carrier that will be used
+        carrier = self.partner_id.purchase_delivery_carrier_id
```
We need a carrier on the PO as initially stated. Use cases showed that the vendor will not always ship with the same carrier. You can keep partner `purchase_delivery_carrier_id` as a fallback value but ideally we shouldn't configure any partner. We are generating drop-shipping PO and the PO carrier should come from the SO delivery carrier.
I would define a field on the PO like `vendor_label_carrier_id`.
This new field should be filled in when the PO is created/updated (done outside this module).
```suggestion
        carrier = self.vendor_label_carrier_id or self.partner_id.purchase_delivery_carrier_id
```

jbaudoux commented on delivery_purchase_label/models/purchase_order.py:
```diff
@@ -0,0 +1,132 @@
+        # Create and process the transer to send the labels
+        values = order._get_purchase_delivery_label_picking_value(carrier)
+        picking = self.env["stock.picking"].with_user(SUPERUSER_ID).create(values)
+        moves = order.order_line._create_stock_moves(picking)
```
Low priority remark:
Instead of making a dry run with `_is_picking_label_uptodate`, you should create a savepoint, make the picking with the moves, then compare and rollback to savepoint if the moves content is the same.
Without this, you have no warranty your dry-run is doing the right thing as you are calling only a subset of the complete picking creation

mt-software-de replied:
This is exactly what is causing problems now. Without the dry-run you are comparing the ordered products and qties against a maybe different transferred products. The moves are not the same as the ordered for example for products with a kit bom.

TDu replied:
Knowing that the `mrp` module will change the moves when the procurement is being run. This is probably too complex to reproduce here.
Would checking  that the purchase lines have not changed (using the `write_date`) since the current delivery label picking has been generated be a good enough solution ?

TDu replied:
Finally using a savepoint

jbaudoux commented on delivery_purchase_label/models/res_partner.py:
```diff
@@ -0,0 +1,14 @@
+
+class ResPartner(models.Model):
+    _inherit = "res.partner"
+    purchase_delivery_carrier_id = fields.Many2one(
```
As this is used for label and not on the real transfer, I would rename to remove confusion
```suggestion
    purchase_label_carrier_id = fields.Many2one(
```

jbaudoux commented on delivery_purchase_label/models/res_partner.py:
```diff
@@ -0,0 +1,14 @@
+    _inherit = "res.partner"
+    purchase_delivery_carrier_id = fields.Many2one(
+        "delivery.carrier",
+        string="Purchase Delivery Method",
```
```suggestion
        string="Purchase Label Delivery Method",
```

jbaudoux reviewed (changes requested):
Some remarks.
Canceling the PO should cancel the generated picking for label. That's a major issue that needs to be tackled.

jbaudoux commented on delivery_purchase_label/models/purchase_order.py:
```diff
@@ -0,0 +1,151 @@
+        if not self._is_valid_for_vendor_labels():
+            return
+        # Find the carrier that will be used
+        carrier = self.vendor_label_carrier_id
+        if not carrier.purchase_label_picking_type:
+            return
```
Shouldn't this be part of `_is_valid_for_vendor_labels` ?

mt-software-de replied:
yes

TDu replied:
Done

mt-software-de committed:
[IMP] delivery_purchase_label: Add many2one field on delivery label transfer to purchase order

TDu committed:
[IMP] delivery_purchase_label: use savepoint

Limit the generation of new transfer for delivery label by using a
savepoint, if no there is no changes the transfer will not be kept

mt-software-de reviewed (approved):
LG. Just the minor remark of @jbaudoux https://github.com/OCA/delivery-carrier/pull/863#discussion_r1773139156

jbaudoux reviewed (approved):
Can you squash fixups ?

jbaudoux commented:
/ocabot merge nobump
